// mongodb-memory-server-core/lib/util/getos/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var fs_1 = require("fs");
var os_1 = require("os");
var child_process_1 = require("child_process");
var util_1 = require("util");
/** Collection of Regexes for "lsb_release -a" parsing */
var LSBRegex = {
    name: /^distributor id:\s*(.*)$/im,
    codename: /^codename:\s*(.*)$/im,
    release: /^release:\s*(.*)$/im,
};
/** Collection of Regexes for "/etc/os-release" parsing */
var OSRegex = {
    name: /^id\s*=\s*"?(.*)"?$/im,
    /** uses VERSION_CODENAME */
    codename: /^version_codename\s*=\s*(.*)$/im,
    release: /^version_id\s*=\s*"?(.*)"?$/im,
};
/**
 * Check if the OS is a LinuxOS Typeguard
 * @param os The OS object to check for
 */
function isLinuxOS(os) {
    return os.os === 'linux';
}
exports.isLinuxOS = isLinuxOS;
/** Get an OS object */
function getOS() {
    return __awaiter(this, void 0, void 0, function () {
        var osName;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    osName = os_1.platform();
                    if (!(osName === 'linux')) return [3 /*break*/, 2];
                    return [4 /*yield*/, getLinuxInfomation()];
                case 1: return [2 /*return*/, _a.sent()];
                case 2: return [2 /*return*/, { os: osName }];
            }
        });
    });
}
exports.default = getOS;
/** Function to outsource Linux Infomation Parsing */
function getLinuxInfomation() {
    var _a;
    return __awaiter(this, void 0, void 0, function () {
        var lsb, err_1, os, err_2, file, os;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    _b.trys.push([0, 2, , 12]);
                    return [4 /*yield*/, util_1.promisify(child_process_1.exec)('lsb_release -a')];
                case 1:
                    lsb = _b.sent();
                    return [2 /*return*/, parseLSB(lsb.stdout)];
                case 2:
                    err_1 = _b.sent();
                    if (!err_1.message.match(/: not found/im)) return [3 /*break*/, 10];
                    _b.label = 3;
                case 3:
                    _b.trys.push([3, 5, , 9]);
                    return [4 /*yield*/, util_1.promisify(fs_1.readFile)('/etc/os-release')];
                case 4:
                    os = _b.sent();
                    return [2 /*return*/, parseOS(os.toString())];
                case 5:
                    err_2 = _b.sent();
                    if (!(((_a = err_2) === null || _a === void 0 ? void 0 : _a.code) === 'ENOENT')) return [3 /*break*/, 8];
                    return [4 /*yield*/, util_1.promisify(fs_1.readdir)('/etc')];
                case 6:
                    file = (_b.sent()).filter(function (v) {
                        return v.match(/.*-release$/im);
                    })[0];
                    if (util_1.isNullOrUndefined(file) || file.length <= 0) {
                        throw new Error('No release file found!');
                    }
                    return [4 /*yield*/, util_1.promisify(child_process_1.exec)('cat /etc/' + file)];
                case 7:
                    os = _b.sent();
                    return [2 /*return*/, parseOS(os.stdout)];
                case 8: throw err_2;
                case 9: return [3 /*break*/, 11];
                case 10: 
                // throw the error in case it is not a "not found" error
                throw err_1;
                case 11: return [3 /*break*/, 12];
                case 12: return [2 /*return*/];
            }
        });
    });
}
/** Function to outsource "lsb_release -a" parsing */
function parseLSB(input) {
    var _a, _b, _c, _d, _e;
    return {
        os: 'linux',
        dist: (_b = (_a = input.match(LSBRegex.name)) === null || _a === void 0 ? void 0 : _a[1], (_b !== null && _b !== void 0 ? _b : 'unkown')),
        codename: (_c = input.match(LSBRegex.codename)) === null || _c === void 0 ? void 0 : _c[1],
        release: (_e = (_d = input.match(LSBRegex.release)) === null || _d === void 0 ? void 0 : _d[1], (_e !== null && _e !== void 0 ? _e : '')),
    };
}
/** Function to outsource "/etc/os-release" parsing */
function parseOS(input) {
    var _a, _b, _c, _d, _e;
    return {
        os: 'linux',
        dist: (_b = (_a = input.match(OSRegex.name)) === null || _a === void 0 ? void 0 : _a[1], (_b !== null && _b !== void 0 ? _b : 'unkown')),
        codename: (_c = input.match(OSRegex.codename)) === null || _c === void 0 ? void 0 : _c[1],
        release: (_e = (_d = input.match(OSRegex.release)) === null || _d === void 0 ? void 0 : _d[1], (_e !== null && _e !== void 0 ? _e : '')),
    };
}
